rbtree-xarray.md

```diff
@@ -0,0 +1,108 @@
+We've since learned that the upstream `RBTree` data structure is deprecated.  Our understanding is that `RBTree` should
+never be used for any new code, and we should use the `XArray` data structure instead.
+
+`XArray` should be fine for all of our use cases in binder except one - the "range allocator".
+We're not sure what to do for this particular case, and are looking for guidance.  [The C driver uses
+RBTree](https://github.com/torvalds/linux/blob/3f01e9fed8454dcd89727016c3e5b2fbb8f8e50c/drivers/android/binder_alloc.h#L83-L85), 
+which led us down that path in the first place.
+
+## TLDR
+**How should we use `XArray` (or some other data structure that is not deprecated) to address the following scenario?**
+
+# Range Allocator
+Range allocator stores collection of "Descriptors":
+```
+struct Descriptor<T> {
+    size: usize,
+    offset: usize,
+    data: Option<T>,
+    state: DescriptorState // Free, Reserved, or Allocated
+}
+```
+
+## Lookups
+We need to look descriptors up one of two ways:
+1. Find a descriptor with a particular `offset`.  Each offset in the collection is unique.
+2. Find the *smallest* descriptor with a state of `Free` and a `size` greater than or equal to a given `size`.  Multiple descriptors can have the same size.
+
+## Merging
+The other nuance is that neighboring descriptors (based on their `offset`) should never *both* be in a state of `Free`.
+When a descriptor transitions to this state, we check it's neighbors, and merge them together accordingly, e.g.:
+
+`(Reserved, Free, Reserved) -> (Reserved, Free, Free) -> (Reserved, Free)` (2nd and 3rd entries merged)
+
+`(Free, Reserved, Free) -> (Free, Free, Free) -> (Free)` (all 3 entries merged into 1)
+
+# Design
+## Naive Approach
+Prior to the above change, descriptors were stored in a LinkedList, which was obviously not great. It meant O(n)
+lookups everywhere.
+```
+struct RangeAllocator<T> {
+    list: LinkedList<Box<Descriptor<T>>>
+}
+```
+
+## Using RBTree
+We improved this with a combination of two RBTree fields:
+```
+struct RangeAllocator<T> {
+    // all descriptors by offset
+    tree: RBTree<usize, Descriptor<T>>
+    // free descriptors by (size, offset)
+    free_tree: RBTree((usize, usize), ())
+}
+```
+
+RBTree allows storing user defined structs as keys, provided they are ordered.  A tuple of integers naturally satisfies this constraint.
+By storing `(size, offset)`, we're able to provide `O(log(n))` lookup of a "best sized" free descriptor.  Since `XArray` only supports
+integer keys, we see the following options:
+
+# Migration Options
+
+## Option 1: Use XArray<Box<LinkedList<usize>>>
+```
+struct RangeAllocator<T> {
+    descriptors: XArray<Box<Descriptor<T>>>,
+    free_indices: XArray<Box<LinkedList<usize>>>
+}
+```
+
+TODO: Get help explaining why this is bad.  I guess We'd have 3 layers of pointers to an integer which I guess is inefficient? I
```

Comment:
How hard would it be to implement a micro-benchmark module in C (rbtree vs radix)? People usually could understand the problem better if they can reproduce it on their own.

Reply:
We actually did some (very naive) benchmarking comparing `LinkedList` to `RBTree` just using kernel log timestamps.  I'm not able to share this doc globally but just gave you access: https://docs.google.com/document/d/1SRJJj8MzzyjLGfBmqaybe5AOrM_GdPZdPWBYlu3YTcI/edit?resourcekey=0-AhO9TyT3L5vM5sLIOrKMQQ#bookmark=id.hrgqt03we1j2

I assume you mean creating a minimal benchmark as its own module/kunit test/something like that?  That sounds feasible, I'm not sure how much bandwidth I have at the moment to do it though 😢 

Another option is to implement this using our proposed `XArray` abstraction and do the same benchmarking we already did.  Maybe the `XArray<Box<LinkedList<usize>>>>` won't be that bad (@Darksonn is skeptical that this will work).

p.s. - I could also copy that doc to a non-google.com gmail account for public viewing if its helpful, but thought it's a bit long winded for this purpose.